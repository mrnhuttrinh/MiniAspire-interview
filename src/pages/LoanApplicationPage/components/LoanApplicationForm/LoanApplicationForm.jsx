import React from 'react';
import block from 'bem-cn';
import { compose } from 'redux';
import { connect } from 'react-redux';
import {
  Field, reduxForm, getFormValues,
} from 'redux-form';

import { InputField, SelectField } from '../../../../shared/reduxForm';
import {
  required, email, phoneNumber,
  minNumber, maxNumber,
} from '../../../../shared/validators';

import './LoanApplicationForm.scss';

const b = block('loan-application-form');

const zeroNumber = minNumber(0);
const billionNumber = maxNumber(1000000000);

function LoanApplicationForm(props) {
  const { handleSubmit, updateLoanForm } = props;

  const handleFormSubmit = (values) => {
    updateLoanForm(values);
  };

  return (
    <form className={b()} onSubmit={handleSubmit(handleFormSubmit)}>
      <div className="title">Contact Information</div>
      <Field
        validate={[required]}
        options={[
          { key: 'Mr', text: 'Mr'},
          { key: 'Mrs', text: 'Mrs'},
          { key: 'Ms', text: 'Ms'},
        ]}
        component={SelectField}
        name="title"
        label="Title"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="firstName"
        label="First Name"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="lastName"
        label="Last Name"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="dateOfBirth"
        label="Date Of Birth"
        type="date"
      />

      <Field
        validate={[required]}
        options={[
          { key: 'Male', text: 'Male'},
          { key: 'Female', text: 'Female'},
          { key: 'Other', text: 'Other'},
        ]}
        component={SelectField}
        name="material"
        label="Material"
      />

      <Field
        validate={[required, phoneNumber]}
        component={InputField}
        name="phone"
        label="Phone"
      />
      
      <Field
        component={InputField}
        validate={[required, email]}
        name="email"
        label="Email"
      />

      <Field
        component={InputField}
        validate={[required]}
        name="address"
        label="Address"
      />

      <br />
      <hr />
      <div className="title">Loan Information</div>
      <Field
        component={InputField}
        validate={[required, zeroNumber, billionNumber]}
        name="amount"
        label="Amount"
        type="number"
      />

      <Field
        validate={[required]}
        options={[
          { key: 'Weekly', text: 'Weekly'},
          { key: 'Monthly', text: 'Monthly'},
          { key: 'Quarterly', text: 'Quarterly'},
          { key: 'Annually', text: 'Annually'},
        ]}
        component={SelectField}
        name="term"
        label="Term"
      />

      <div>
        <button type="submit" className="submit-button">
          Send Application Now
        </button>
      </div>
    </form>
  );
}

const LOAN_APPLICATION_FORM = 'loanApplicationForm';

const mapStateToProps = (state) => {
  return {
    values: getFormValues(LOAN_APPLICATION_FORM)(state)
  };
};

export { LoanApplicationForm };

export default compose(
  reduxForm({
    form: LOAN_APPLICATION_FORM,
    initialValues: {
      title: 'Mr',
      material: 'Male',
      term: 'Weekly'
    },
  }),
  connect(mapStateToProps),
)(LoanApplicationForm);
