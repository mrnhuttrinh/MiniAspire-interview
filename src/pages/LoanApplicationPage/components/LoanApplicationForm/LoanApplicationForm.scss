.loan-application-form {
  margin: 0 auto 30px;

  width: 600px;
  height: auto;
  background-color: rgba($color: #000000, $alpha: 0.05);
  border-radius: 5px;
  box-shadow: 0 0 5px 0;
  padding: 20px;
  display: flex;
  flex-direction: column;

  .submit-button {
    padding: 5px 15px;
    margin-left: 50%;
    margin-top: 20px;
    transform: translateX(-50%);
    border-radius: 5px;
  }

  .title {
    margin: 10px 0;
    font-weight: bold;
    font-size: 18px;
  }
}