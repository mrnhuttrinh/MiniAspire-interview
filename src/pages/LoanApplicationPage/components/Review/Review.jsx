import React, { useState } from 'react';
import block from 'bem-cn';

import './Review.scss';

const b = block('loan-application-review');

function Review({ loanApplication, updateLoadAmount }) {
  const [repayAmount, setRepayAmount] = useState(0);

  const handleRepayOnChange = (e) => {
    const numberValue = Number(e.target.value);
    setRepayAmount(numberValue);
  };

  const handleRepayButtonClick = () => {
    if (repayAmount > 0 && repayAmount <= Number(loanApplication.amount)) {
      updateLoadAmount(repayAmount);
      setRepayAmount('');
    }
  };

  return (
    <div className={b()}>
      <div className="field-data">
        <label>Title</label>
        <span>{loanApplication.title}</span>
      </div>
      <div className="field-data">
        <label>First Name</label>
        <span>{loanApplication.firstName}</span>
      </div>
      <div className="field-data">
        <label>Last Name</label>
        <span>{loanApplication.lastName}</span>
      </div>
      <div className="field-data">
        <label>Date Of Birth</label>
        <span>{loanApplication.dateOfBirth}</span>
      </div>
      <div className="field-data">
        <label>Phone</label>
        <span>{loanApplication.phone}</span>
      </div>
      <div className="field-data">
        <label>Email</label>
        <span>{loanApplication.email}</span>
      </div>
      <div className="field-data">
        <label>Address</label>
        <span>{loanApplication.address}</span>
      </div>

      <br />
      <hr />
      <div className="field-data">
        <label>Amount</label>
        <span>{loanApplication.amount}</span>
      </div>

      <div className="field-data">
        <label>Term</label>
        <span>{loanApplication.term}</span>
      </div>

      <br />
      <hr />

      <div className="title">Repayment</div>
      {
        loanApplication.amount ? (
          <>
            <div className="field">
              <label className="label">Repayment amount</label>
              <div className={'input'}>
                <input
                  className="input--control"
                  placeholder="Repayment amount"
                  value={repayAmount}
                  onChange={handleRepayOnChange}
                  type="number"
                />
                {repayAmount < 0 && <span className="input--error">Repayment Amount must be greater than 0</span>}
                {repayAmount > Number(loanApplication.amount) && <span className="input--error">Repayment Amount must be less than {loanApplication.amount}</span>}
              </div>
            </div>

            <div>
              <button
                type="submit"
                className="submit-button"
                disabled={repayAmount < 0 || repayAmount > Number(loanApplication.amount)}
                onClick={handleRepayButtonClick}
              >
                Repayment
              </button>
            </div>
          </>
        ) : 'You have payment successufully'
      }
    </div>
  );
}

export default Review;
