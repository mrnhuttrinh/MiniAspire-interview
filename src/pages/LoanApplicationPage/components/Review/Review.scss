.loan-application-review {
  margin: 0 auto 30px;

  width: 600px;
  height: auto;
  background-color: rgba($color: #000000, $alpha: 0.05);
  border-radius: 5px;
  box-shadow: 0 0 5px 0;
  padding: 20px;
  display: flex;
  flex-direction: column;

  .field-data {
    width: 100%;
    display: flex;
    align-items: center;
    margin: 5px 0;

    > label {
      font-weight: bold;
      flex: 0 150px;
    }

    > span {
      height: auto;
      width: 100%;
      font-size: 20px;
      padding: 5px;
    }
  }

  .submit-button {
    padding: 5px 15px;
    margin-left: 50%;
    margin-top: 20px;
    transform: translateX(-50%);
    border-radius: 5px;
  }

  .title {
    margin: 10px 0;
    font-weight: bold;
    font-size: 18px;
  }
}