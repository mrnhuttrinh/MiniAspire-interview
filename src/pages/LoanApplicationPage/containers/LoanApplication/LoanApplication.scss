.loan-application {
  $r: &;

  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;

  .loan-application__heading {
    flex: 0 50px;
    display: flex;
    align-items: center;
    justify-content: center;
    font-weight: bold;
    font-size: 20px;
  }
}