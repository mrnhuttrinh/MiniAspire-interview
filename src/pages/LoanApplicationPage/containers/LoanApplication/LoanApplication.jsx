import React from 'react';
import block from 'bem-cn';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import _ from 'lodash';

import * as actions from '../../actions';
import { selectedLoanData } from '../../reducers/selectors';
import './LoanApplication.scss';

import Review from '../../components/Review/Review';
import LoanApplicationForm from '../../components/LoanApplicationForm/LoanApplicationForm';

const b = block('loan-application');

function LoanApplication(props) {
  const {
    updateLoanApplicationForm, loanApplication,
    updateLoadAmount,
  } = props;

  const updateLoanForm = (values) => {
    updateLoanApplicationForm(values);
  };

  if (!_.isEmpty(loanApplication)) {
    return (
      <div className={b()}>
        <div className={b('heading')}>Review Application</div>
        <Review
          loanApplication={loanApplication}
          updateLoadAmount={updateLoadAmount}
        />
      </div>
    );
  }

  return (
    <div className={b()}>
      <div className={b('heading')}>Loan Application Form</div>
      <LoanApplicationForm
        updateLoanForm={updateLoanForm}
      />
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    loanApplication: selectedLoanData(state),
  };
};

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(actions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(LoanApplication);
