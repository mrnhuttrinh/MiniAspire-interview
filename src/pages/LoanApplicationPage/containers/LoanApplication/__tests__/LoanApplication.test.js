import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import LoanApplication, { LoanApplication as LoanApplicationComponent} from '../LoanApplication';
import LoanApplicationForm from '../../../components/LoanApplicationForm/LoanApplicationForm';
import Review from '../../../components/Review/Review';

const buildStore = configureStore();

describe('LoanApplication', () => {
  let store;
  let wrapper;

  it('should render LoanApplication with child LoanApplicationForm Component', () => {
    const initialState = { loanApplication: null };
    store = buildStore(initialState);
    wrapper = shallow(
      <Provider store={store}>
        <LoanApplication />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render LoanApplication with child LoanApplicationForm Component', () => {
    wrapper = shallow(<LoanApplicationComponent loanApplication={null} />);
    expect(wrapper.find(LoanApplicationForm)).toHaveLength(1);
  });

  it('should render LoanApplication with child Review Component', () => {
    const loanApplication = {
      firstName: 'John'
    };
    wrapper = shallow(<LoanApplicationComponent loanApplication={loanApplication} />);
    expect(wrapper.find(Review)).toHaveLength(1);
  });
});
