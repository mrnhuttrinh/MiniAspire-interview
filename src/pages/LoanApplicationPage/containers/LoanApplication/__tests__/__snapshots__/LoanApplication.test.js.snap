// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanApplication should render LoanApplication with child LoanApplicationForm Component 1`] = `<Connect(LoanApplication) />`;
